src/tokenizer/shingle_filter.rs

```diff
@@ -0,0 +1,207 @@
+use super::{BoxTokenStream, Token, TokenFilter, TokenStream};
+
+#[derive(Clone)]
+pub struct FixedShingleFilter {
+    size: usize,
+    separator: char,
+}
+
+impl FixedShingleFilter {
+    pub fn new(size: usize) -> Self {
+        FixedShingleFilter::new_with_separator(size, '_')
+    }
+
+    pub fn new_with_separator(size: usize, separator: char) -> Self {
+        FixedShingleFilter { size, separator }
+    }
+}
+
+impl TokenFilter for FixedShingleFilter {
+    fn transform<'a>(&self, token_stream: BoxTokenStream<'a>) -> BoxTokenStream<'a> {
+        BoxTokenStream::from(ShingleFilterStream {
+            size: self.size,
+            separator: self.separator,
+            buffer: String::with_capacity(100),
+            pendings: Vec::with_capacity(self.size - 1),
+            tail: token_stream,
+            original_token_position: 0,
+        })
+    }
+}
+
+pub struct ShingleFilterStream<'a> {
+    size: usize,
+    separator: char,
+    buffer: String,
+    pendings: Vec<Token>,
+    tail: BoxTokenStream<'a>,
+    original_token_position: usize,
+}
+
+impl<'a> ShingleFilterStream<'a> {
+    fn is_window_filled(&self) -> bool {
+        self.pendings.len() == self.size - 1
+    }
+
+    fn push_token(&mut self, mut token: Token) {
+        if self.is_window_filled() {
+            self.pendings.remove(0);
+        }
+        token.position = self.original_token_position;
+        self.pendings.push(token);
+    }
+}
+
+fn make_shingle(token: &mut Token, pendings: &Vec<Token>, buffer: &mut String, separator: char) {
+    buffer.clear();
+    buffer.push_str(token.text.as_str());
+
+    // set text
+    token.text.clear();
+    for pend_token in pendings {
+        token.text.push_str(pend_token.text.as_str());
+        token.text.push(separator);
+    }
+    token.text.push_str(buffer.as_str());
+
+    // set offset and position
+    if let Some(first_token) = pendings.get(0) {
+        token.offset_from = first_token.offset_from;
+        token.position = first_token.position;
+    }
+}
+
+impl<'a> TokenStream for ShingleFilterStream<'a> {
+    fn advance(&mut self) -> bool {
+        while !self.is_window_filled() {
+            if self.tail.advance() {
+                self.push_token(self.tail.token().clone());
+                self.original_token_position += 1;  // This is not compatible with StopWordFilter
```

Comment:
Can you comment in the code the reason why you chose not to use `self.tail.token().position` for this?
I assume this has to do with matching some phrase

Reply:
Thanks for your suggestion. Actually, I still don't have a clear idea about position handling in tantivy; this silly hack should be replaced with more sensible code. I'll try to find the right way for this.

Reply:
I would not be surprise if we are inconsistent on positions. I would advise you to be very critical on tantivy approach, and I am ok to adopt the lucene way in all tokenizer.

My understanding is that for something like a stop word filter, depending on how we deal with position we will have a different outcome for the following query / doc pair.
- query: the cheapest on internet 
- doc: the cheapest on the internet

In other words, should queries act as 
`the (STOP WORDS)* cheapest (STOP WORDS)* internet`
or 
`the cheapest (STOP WORDS){2} internet`

There is a decent case for both. The best is probably to adopt the lucene way and stay consistent with all TokenFilter

Reply:
@mocobeta Do you want help to dig into what should be the right choice for position here?

Reply:
@fulmicoton I can't find the time right now, but it could be good homework for the coming winter vacation.